File: daily-exercise/special_topic/array/minimum-size-subarray-sum.py
# -*- coding: utf-8 -*-
# @Time    : 2022/8/11 19:58
# @Author  : kevin
# @Version : python 3.7
# @Desc    : minimum-size-subarray-sum
# 长度最小的子数组：滑动窗口

class Solution:
    def minSubArrayLen(self, target, nums):
        n = len(nums)
        if not nums or n == 0:
            return 0
        slow = 0
        lengthMin = len(nums) + 1
        numSum = 0
        for fast in range(n):
            numSum += nums[fast]
            while numSum >= target:
                lengthMin = min(lengthMin, fast - slow + 1)
                numSum -= nums[slow]
                slow += 1
        if lengthMin == len(nums) + 1:
            return 0
        return lengthMin


test = Solution()
l = test.minSubArrayLen(7, [2, 3, 1, 2, 4, 3])
print(l)

File: daily-exercise/special_topic/array/minimum-size-subarray-sum.go
func minSubArrayLen(target int, nums []int) int {
    n := len(nums)
    if nums == nil || n == 0{
        return 0
    }
    slow,sumNum := 0,0
    minLength := n+1
    for fast := 0;fast<n;fast++{
        sumNum += nums[fast]
        for sumNum >= target{
            minLength = min(minLength,fast-slow+1)
            sumNum -= nums[slow]
            slow++
        }
    }
    if minLength == n+1{
        return 0
    }else{
        return minLength
    }
}

func min(x,y int) int {
    if x > y{
        return y
    }
    return x
}
